### ISSUE TITLE
Run code (not just SQL) at connection initialization

### ISSUE THREAD

**Issue by gotson** [brettwooldridge/HikariCP#1812]
Hi,

i see there is `connectionInitSql` to run some SQL on each new connection.

I am using SQLite, and in order to use the `REGEXP` operator, a user defined function must be registered for each new connection, [like so](https://stackoverflow.com/a/28587200/2041863).

At the moment i'm adding this at application startup, by unwrapping the connection provided by HikariCP. The problem is that if HikariCP loses the connection, and recreates one, the user defined function is not there anymore.

Would it be possible to run Java code, with some kind of hooks or events or lambda, for every new connection?

Thanks!

**Comment by gotson** [comment]
I managed to get around with the same solution as https://github.com/brettwooldridge/HikariCP/issues/523#issuecomment-165603577

Here is my working solution for those interested.

```kotlin
class SqliteUdfDataSource : SimpleDriverDataSource() {

  override fun getConnection(): Connection =
    super.getConnection().also { createUdfRegexp(it as SQLiteConnection) }

  override fun getConnection(username: String, password: String): Connection =
    super.getConnection(username, password).also { createUdfRegexp(it as SQLiteConnection) }

  private fun createUdfRegexp(connection: SQLiteConnection) {
    Function.create(
      connection, "REGEXP",
      object : Function() {
        override fun xFunc() {
          val regexp = (value_text(0) ?: "").toRegex(RegexOption.IGNORE_CASE)
          val text = value_text(1) ?: ""

          result(if (regexp.containsMatchIn(text)) 1 else 0)
        }
      }
    )
  }
}
```

Then initialize the Hikari datasource like this:

```kotlin
fun sqliteDataSource(): DataSource {

    val sqliteUdfDataSource = DataSourceBuilder.create()
      .driverClassName("org.sqlite.JDBC")
      .url("jdbc:sqlite:${komgaProperties.database.file}?foreign_keys=on;")
      .type(SqliteUdfDataSource::class.java)
      .build()

    return HikariDataSource(
      HikariConfig().apply {
        dataSource = sqliteUdfDataSource
        poolName = "SqliteUdfPool"
        maximumPoolSize = 1
      })
  }
```

**Comment by allgit1** [comment]
I encountered the same problem while migrating old code. The database connection pool in the old code used a sessionId queried from the database to uniquely identify a connection for logging purposes. I tried to add a hook in HikariCP at the point where a connection is initially created from the database. However, I discovered that HikariCP does not provide such an entry point. The only available override is for the getConnection method, which is for retrieving a connection from the pool. This would be too costly for me. According to some comments in the issues section, the author is not willing to add this entry point. Thus, I eventually had to modify the old code's approach. I am leaving this message to save time for others who might be looking for similar issues and solutions in HikariCP.
